### ISSUE TITLE
Enhancing Actions Autoscaling Capabilities

### ISSUE THREAD

**Issue by cobak78** [go-gitea/gitea#32862]
### Feature Description

Hello,

I wanted to share an idea we're exploring to improve the scalability and efficiency of Actions runners, using KEDA in a kubernetes environment, gather your thoughts and eventually create a PR to add the needed endpoints to achieve this feature.

This idea is well explainen here on Forgejo project. https://codeberg.org/forgejo/discussions/issues/241

According to that my feature request is to create three new enpoints on Gitea admin path to allow getting jobs to be executed in a previous registered runner.

This wil be the swagger spec of the global register endpoint 
```
	// swagger:operation GET /admin/runners/actions/jobs admin adminSearchRunJobs
	// ---
	// summary: Search action jobs according filter conditions
	// produces:
	// - application/json
	// parameters:
	// - name: labels
	//   in: query
	//   description: a comma separated list of run job labels to search for
	//   type: string
	// responses:
	//   "200":
	//     "$ref": "#/responses/RunJobList"
	//   "403":
	//     "$ref": "#/responses/forbidden"
```

This will allow a KEDA Autoscaler to get how many runner should spawn according to the number of jobs for an specific label. Run the a scaledJob with the configured runner previously registered, and close it when its done.

Please let me know if you need more information or to show more about our working POC.

Thanks

### Screenshots

![Screenshot 2024-12-16 at 17 02 35](https://github.com/user-attachments/assets/e2338f01-4e2f-4c7b-bd53-94f7240fb4c7)
<img width="453" alt="Screenshot 2024-12-16 at 17 04 37" src="https://github.com/user-attachments/assets/da1bb191-328b-42bf-adee-6d44347fdd8d" />

**Comment by lunny** [comment]
The design will require creating an admin token and it will be suitable for internal Gitea instances. However, for a public instance, it will not allow every organization to use its own auto scaler runners.

Other questions about the whole design. I don't know KEDA, if I'm wrong please let me know.
How to deploy the runners with the configuration.
1 minimal idle runner number
2 maximal running runners number
3 the max memory the runners can use
4 the max CPU the runners can use
5  how to isolate one repository's tasks from others to avoid possible access cross-tasks

**Comment by cobak78** [comment]
Thank you for the comment. Addresing your questions:
1. The minimum number of iddle runners is defined by `minReplicaCount' and can be 0, as is the KEDA autoscaler now how makes request to the endpoint to know when to spawn a new runner.
2. the maximum number of concurrent runners i defined by the `maxReplicaCount` parameter in the configuration.
3 & 4. The memory and cpu limits can be specified directly in teh autoscaler configuration, setting as in any other container spec.
5. Runner registration should be done prior to enabling the autoscaler. Then there will be three types of endpoints based on access levels: repository, user, and global. Just as a different token is generated for each level and job queue, the registred runner spawned by KEDA will correspond to a specific queue, ensuring they remain isolated form on another when is on user or repo level.

**Comment by cobak78** [comment]
you are right, we only need idle status jobs, more precisely `actions_model.StatusWaiting` and `actions_model.StatusRunning` as the scaled jobs from KEDA autoscaler needs to keep the job running til it ends. 

There will be a new autoscaler in KEDA looking for his endpoints and mostly inspired in the github KEDA autoscaler https://keda.sh/docs/2.11/scalers/github-runner/

**Comment by Zettat123** [comment]
Thanks for the suggestion. It would be great if Gitea Actions had stronger autoscaling capabilities. You are welcome to send PRs to support this feature!

**Comment by anyidea** [comment]
any progress?

**Comment by bitchecker** [comment]
Hi,
any news on this really interesting feature?
